File: 프로그래머스/Level2/프린터.py
# 코딩테스트 연습 > 스택/큐 > 프린터 (Level2)


def solution(priorities, location):
    answer = 0

    while True:
        max_priority = max(priorities)  # 현재의 최대 우선순위 값

        current = priorities.pop(0)  # 대기목록에서 맨앞에 있는 것을 꺼낸다.

        # 방금 꺼낸 것이 최대의 우선순위를 가지는 경우,
        if current == max_priority:
            answer += 1  # 프린트 횟수 1 증가
            if location == 0:
                break  # 내 문서인 경우 반복문 종료
            else:
                location -= 1  # 내 문서가 아닌 경우 내 문서 위치 1만큼 감소
        # 방금 꺼낸 것이 최대 우선순위가 아닌 경우,
        else:
            priorities.append(current)  # 대기목록에 다시 넣는다.
            if location == 0:
                location = len(priorities) - 1  # 내 문서인 경우, 위치 갱신
            else:
                location -= 1  # 내 문서가 아닌 경우, 내 문서 위치 1만큼 감소

    return answer


print(solution([1, 1, 9, 1, 1, 1], 0))  # 5

File: 프로그래머스/Level2/Javascript/프린터.js
function solution(priorities, location) {
    let answer = 0;
    
    while (true) {
        const maxPriority = Math.max.apply(null, priorities);
        const current = priorities.shift();

        // 맨 앞 문서의 우선순위가 최대값과 같은 경우, 인쇄 처리한다.
        if (current === maxPriority) {
            answer++;
            // 방금 인쇄한 문서가 내가 요청한 것이었다면, 반복문을 중단한다.
            if (location === 0) {
                break;
            }
        } 
        // 맨 앞 문서의 우선순위가 최대값보다 작은 경우, Queue의 맨 뒤로 보낸다.
        else {
            priorities.push(current);
        }

        // 문서 위치를 갱신한다.
        location = location === 0 ? priorities.length - 1 : location - 1;
    }
    
    return answer;
}